In [1]:
#Penglab (Abuse of Google Colab for fun and profit)
#by mxrch

#Choose what you want to install
hashcat = True
john = False
hydra = False

#Which shell ? (only one, or none)
#⚠️ Don't select a shell to run bash commands in the Google Colab way (you'll see). Stable.
ssh = False #It uses ngrok. Stable and interactive, but take a little time to setup
python_shell = True #Useful for little hashes, but not for long hours, since you'll see the output only when it finish. Not interactive, but stable.
bash_shell = False #Useful for big hashes. but you'll don't see your input (surely a Google Colab protection). Stable.

#Wordlists (see weakpass.com to download them)
wordlists_dir = "wordlists"

rockyou = True #For CTFs, especially Hack The Box (133.44 Mb)
hashesorg2019 = False #Very heavy (12.79 Gb) but has a good rate, if you're determinated

#---------------------------

if (python_shell and bash_shell) or (bash_shell and ssh) or (ssh and python_shell) :
    print("Please do a choice")
    exit()

In [2]:
#Wordlists
import os

os.system("wordlists_dir={}".format(wordlists_dir))
!mkdir ./$wordlists_dir
if rockyou:
    !printf "[+] Downloading the Rockyou wordlist...\n"
    !cd $wordlists_dir && wget https://download.weakpass.com/wordlists/90/rockyou.txt.gz
    !printf "[+] Wordlist downloaded !\n[+] Extraction...\n"
    !cd $wordlists_dir && gunzip rockyou.txt.gz
    !printf "[+] Finished !\n[+] Location : $(pwd)/$wordlists_dir/$(ls wordlists | grep rockyou)"

if hashesorg2019:
    !printf "[+] Downloading the HashesOrg2019 wordlist...\n"
    !cd $wordlists_dir && wget https://download.weakpass.com/wordlists/1851/hashesorg2019.gz
    !printf "[+] Wordlist downloaded !\n[+] Extraction...\n"
    !cd $wordlists_dir && gunzip hashesorg2019.gz
    !printf "[+] Finished !\n[+] Location : $(pwd)/$wordlists_dir/$(ls wordlists | grep hashesorg2019)"

[+] Downloading the Rockyou wordlist...
--2025-11-27 12:27:43--  https://download.weakpass.com/wordlists/90/rockyou.txt.gz
Resolving download.weakpass.com (download.weakpass.com)... 172.67.206.163, 104.21.22.190, 2606:4700:3036::6815:16be, ...
Connecting to download.weakpass.com (download.weakpass.com)|172.67.206.163|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 53357062 (51M) [application/octet-stream]
Saving to: ‘rockyou.txt.gz’

rockyou.txt.gz      100%[===================>]  50.88M   127KB/s    in 5m 35s  

2025-11-27 12:33:18 (155 KB/s) - ‘rockyou.txt.gz’ saved [53357062/53357062]

[+] Wordlist downloaded !
[+] Extraction...
[+] Finished !
[+] Location : /content/wordlists/rockyou.txt

In [ ]:
hashcat = True
#Install of hashcat
if hashcat:
    print("Installation of hashcat...")
    !apt install cmake build-essential -y
    !apt install checkinstall git -y
    !git clone https://github.com/hashcat/hashcat.git && cd hashcat && make -j 8 && make install

Installation of hashcat...
Reading package lists... Done
Building dependency tree... Done
Reading state information... Done
build-essential is already the newest version (12.9ubuntu3).
cmake is already the newest version (3.22.1-1ubuntu1.22.04.2).
0 upgraded, 0 newly installed, 0 to remove and 41 not upgraded.
Reading package lists... Done
Building dependency tree... Done
Reading state information... Done
git is already the newest version (1:2.34.1-1ubuntu1.15).
Suggested packages:
  gettext
The following NEW packages will be installed:
  checkinstall
0 upgraded, 1 newly installed, 0 to remove and 41 not upgraded.
Need to get 107 kB of archives.
After this operation, 446 kB of additional disk space will be used.
Get:1 http://archive.ubuntu.com/ubuntu jammy/universe amd64 checkinstall amd64 1.6.2+git20170426.d24a630-2ubuntu2 [107 kB]
Fetched 107 kB in 1s (144 kB/s)
Selecting previously unselected package checkinstall.
(Reading database ... 121713 files and directories currently installe

In [1]:
# List the contents of the hashcat directory to find the executable
!ls -R /content/hashcat/

ls: cannot access '/content/hashcat/': No such file or directory


In [ ]:
# List the contents of the hashcat directory to find the executable
!ls -R /content/hashcat/

In [ ]:
!/usr/local/bin/hashcat -h

In [4]:
# Update PATH to include Cargo's bin directory
!bash -c "source $HOME/.cargo/env && echo 'PATH updated successfully.'"


bash: line 1: /root/.cargo/env: No such file or directory


In [1]:
# Install Rust
print("Installing Rust...")
!curl --proto '=https' --tlsv1.2 -sSf https://sh.rustup.rs | sh -s -- -y
# Add cargo to PATH for current session
!source $HOME/.cargo/env
print("Rust installation complete.")

Installing Rust...
info: downloading installer
info: profile set to 'default'
info: default host triple is x86_64-unknown-linux-gnu
info: syncing channel updates for 'stable-x86_64-unknown-linux-gnu'
info: latest update on 2025-11-10, rust version 1.91.1 (ed61e7d7e 2025-11-07)
info: downloading component 'cargo'
info: downloading component 'clippy'
info: downloading component 'rust-docs'
info: downloading component 'rust-std'
info: downloading component 'rustc'
info: downloading component 'rustfmt'
info: installing component 'cargo'
 10.2 MiB /  10.2 MiB (100 %)   9.2 MiB/s in  1s
info: installing component 'clippy'
info: installing component 'rust-docs'
 20.4 MiB /  20.4 MiB (100 %)   1.7 MiB/s in  9s
info: installing component 'rust-std'
 27.9 MiB /  27.9 MiB (100 %)   8.0 MiB/s in  3s
info: installing component 'rustc'
 74.5 MiB /  74.5 MiB (100 %)   6.7 MiB/s in  9s
info: installing component 'rustfmt'
info: default toolchain set to 'stable-x86_64-unknown-linux-gnu'

  stable-x86_6

In [ ]:
#Install of john
if john:
    print("Installation of john...")
    !apt-get install john -y

In [ ]:
#Install of hydra
if hydra:
    print("Installation of hydra...")
    !apt install cmake build-essential -y
    !apt install checkinstall git -y
    !git clone https://github.com/vanhauser-thc/thc-hydra.git && cd thc-hydra && ./configure && make && make install

In [ ]:
#Setting up shells
from termcolor import colored
import os

if ssh:
    print("Setting up SSH...")
    !pip install git+https://github.com/demotomohiro/remocolab.git
    import remocolab
    remocolab.setupSSHD()

elif python_shell:
    os.system("touch /tmp/cmd && touch /tmp/status")

    template_cmd = "echo -n $(whoami)[$(hostname)[$(pwd) &> /tmp/status"
    os.system("bash -c '{}'".format(template_cmd))
    output = {"cmd": "", "status": ""}
    with open('/tmp/cmd', 'r') as f:
        output["cmd"] = f.read()
    with open('/tmp/status', 'r') as f:
        output["status"] = f.read()
    prefixes = output["status"].split("[")
    path = prefixes[2].replace('\n', '')
    prefix = colored(prefixes[0] + "@" + prefixes[1], "red") + ":" + colored(path, "cyan") + "$ "
    print("")

    while 1:
        print(prefix, end='')
        cmd = input()
        template_cmd = "cd {} && ".format(path) + "" + cmd + " &> /tmp/cmd ; echo $(whoami)[$(hostname)[$(pwd) &> /tmp/status"
        os.system("bash -c '{}'".format(template_cmd))
        output = {"cmd": "", "status": ""}
        with open('/tmp/cmd', 'r') as f:
            output["cmd"] = f.read()
        with open('/tmp/status', 'r') as f:
            output["status"] = f.read()
        prefixes = output["status"].split("[")
        path = prefixes[2].replace('\n', '')
        prefix = colored(prefixes[0] + "@" + prefixes[1], "red") + ":" + colored(path, "cyan") + "$ "
        print(output["cmd"])

elif bash_shell:
    !/bin/bash

else:
    print('\n🚨🚨🚨\n🎉 Your environment is ready !\nWant to run it from the code blocks ?')
    print('Just use the code block below and start your commands by a "!" (make sure to uncomment them).')
    print('▶️ When you\'re ready, hit the run button at the left of the code block !\n')
    print('|\n|\n|\nV')

In [ ]:
# Put your code here :
# !hashcat -h
# !john --help

# Task
Hashcat'i kurmak için önce mevcut `/content/hashcat` dizinini sileceğiz, ardından Hashcat'in en son önceden derlenmiş sürümünü resmi GitHub yayınlarından indirecek, içeriği çıkaracak, hashcat yürütülebilir dosyasının tam yolunu bulacak ve kurulumu `hashcat -h` komutuyla doğrulayacağız.

## Mevcut Hashcat Dizinini Temizle

### Subtask:
Herhangi bir önceki başarısız kurulumun kalıntılarını kaldırmak için `/content/hashcat` dizinini sil.


**Reasoning**:
To remove the `/content/hashcat` directory and its contents, I will use the `rm -rf` command.



In [1]:
!rm -rf /content/hashcat

## Hashcat İndir

### Subtask:
Hashcat'in en son önceden derlenmiş sürümünü resmi GitHub yayınlarından indireceğiz.


**Reasoning**:
The subtask requires downloading the latest pre-compiled version of Hashcat. I will use the `wget` command to download the `hashcat-6.2.6.tar.gz` archive, which is a common stable release, from the official GitHub releases page as per the instructions.



In [2]:
!
LATEST_HASHCAT_VERSION="6.2.6"
!wget https://github.com/hashcat/hashcat/releases/download/v${LATEST_HASHCAT_VERSION}/hashcat-${LATEST_HASHCAT_VERSION}.tar.gz

--2025-11-27 13:12:10--  https://github.com/hashcat/hashcat/releases/download/v.2.6/hashcat-.2.6.tar.gz
Resolving github.com (github.com)... 140.82.113.4
Connecting to github.com (github.com)|140.82.113.4|:443... connected.
HTTP request sent, awaiting response... 404 Not Found
2025-11-27 13:12:10 ERROR 404: Not Found.



**Reasoning**:
The previous attempt to download Hashcat failed because the URL was malformed due to an issue with variable interpolation in the `!` magic command. I will reconstruct the `wget` command by explicitly formatting the URL with the `LATEST_HASHCAT_VERSION` variable using a Python f-string to ensure correct expansion before executing the shell command.



In [3]:
LATEST_HASHCAT_VERSION = "6.2.6"
url = f"https://github.com/hashcat/hashcat/releases/download/v{LATEST_HASHCAT_VERSION}/hashcat-{LATEST_HASHCAT_VERSION}.tar.gz"
!wget {url}

--2025-11-27 13:12:32--  https://github.com/hashcat/hashcat/releases/download/v6.2.6/hashcat-6.2.6.tar.gz
Resolving github.com (github.com)... 140.82.114.4
Connecting to github.com (github.com)|140.82.114.4|:443... connected.
HTTP request sent, awaiting response... 404 Not Found
2025-11-27 13:12:33 ERROR 404: Not Found.



**Reasoning**:
The previous attempt to download Hashcat failed because `hashcat-6.2.6.tar.gz` is not available as a pre-compiled binary on the official GitHub releases page. Hashcat provides pre-compiled Linux binaries in `.7z` format. To proceed with downloading a pre-compiled version, I will download `hashcat-6.2.6.7z` instead, which is the correct pre-compiled binary for this version.



In [5]:
LATEST_HASHCAT_VERSION = "6.2.6"
url = f"https://github.com/hashcat/hashcat/releases/download/v{LATEST_HASHCAT_VERSION}/hashcat-{LATEST_HASHCAT_VERSION}.7z"
print(f"Attempting to download: {url}")
!wget {url}

Attempting to download: https://github.com/hashcat/hashcat/releases/download/v6.2.6/hashcat-6.2.6.7z
--2025-11-27 13:13:32--  https://github.com/hashcat/hashcat/releases/download/v6.2.6/hashcat-6.2.6.7z
Resolving github.com (github.com)... 140.82.112.3
Connecting to github.com (github.com)|140.82.112.3|:443... connected.
HTTP request sent, awaiting response... 302 Found
Location: https://release-assets.githubusercontent.com/github-production-release-asset/47409672/ea6aa4ab-f9ec-4c03-83e1-dcfaf0b24ad2?sp=r&sv=2018-11-09&sr=b&spr=https&se=2025-11-27T13%3A55%3A20Z&rscd=attachment%3B+filename%3Dhashcat-6.2.6.7z&rsct=application%2Foctet-stream&skoid=96c2d410-5711-43a1-aedd-ab1947aa7ab0&sktid=398a6654-997b-47e9-b12b-9515b896b4de&skt=2025-11-27T12%3A54%3A29Z&ske=2025-11-27T13%3A55%3A20Z&sks=b&skv=2018-11-09&sig=Uv%2By302u4p1Q404M9A4hbngl91MbuX6UQSSJ%2FzLQd%2B0%3D&jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmVsZWFzZS1hc3NldHMuZ2l0aHVidXNlcmNvbnRlbnQuY29tIiwia2

**Reasoning**:
I have successfully downloaded the Hashcat archive in `.7z` format. To extract its contents, I need to use the `7z` command. Therefore, I will first install the `p7zip-full` package, which provides the `7z` utility.



In [6]:
print("Installing p7zip-full for .7z extraction...")
!apt-get update
!apt-get install p7zip-full -y

Installing p7zip-full for .7z extraction...
Get:1 http://security.ubuntu.com/ubuntu jammy-security InRelease [129 kB]
Get:2 https://cloud.r-project.org/bin/linux/ubuntu jammy-cran40/ InRelease [3,632 B]
Hit:3 https://cli.github.com/packages stable InRelease
Get:4 https://developer.download.nvidia.com/compute/cuda/repos/ubuntu2204/x86_64  InRelease [1,581 B]
Hit:5 http://archive.ubuntu.com/ubuntu jammy InRelease
Get:6 https://r2u.stat.illinois.edu/ubuntu jammy InRelease [6,555 B]
Get:7 http://archive.ubuntu.com/ubuntu jammy-updates InRelease [128 kB]
Get:8 https://developer.download.nvidia.com/compute/cuda/repos/ubuntu2204/x86_64  Packages [2,153 kB]
Get:9 https://ppa.launchpadcontent.net/deadsnakes/ppa/ubuntu jammy InRelease [18.1 kB]
Get:10 http://security.ubuntu.com/ubuntu jammy-security/main amd64 Packages [3,539 kB]
Hit:11 https://ppa.launchpadcontent.net/graphics-drivers/ppa/ubuntu jammy InRelease
Get:12 http://archive.ubuntu.com/ubuntu jammy-backports InRelease [127 kB]
Hit:13 ht

**Reasoning**:
Now that `p7zip-full` is installed and the Hashcat `.7z` archive has been downloaded, I will extract its contents using the `7z x` command.



In [7]:
LATEST_HASHCAT_VERSION = "6.2.6"
archive_name = f"hashcat-{LATEST_HASHCAT_VERSION}.7z"
print(f"Extracting {archive_name}...")
!7z x {archive_name}

Extracting hashcat-6.2.6.7z...

7-Zip [64] 16.02 : Copyright (c) 1999-2016 Igor Pavlov : 2016-05-21
p7zip Version 16.02 (locale=en_US.UTF-8,Utf16=on,HugeFiles=on,64 bits,2 CPUs Intel(R) Xeon(R) CPU @ 2.20GHz (406F0),ASM,AES-NI)

Scanning the drive for archives:
  0M Scan         1 file, 20951515 bytes (20 MiB)

Extracting archive: hashcat-6.2.6.7z
--
Path = hashcat-6.2.6.7z
Type = 7z
Physical Size = 20951515
Headers Size = 19192
Method = LZMA2:384m LZMA:20 BCJ2
Solid = +
Blocks = 2

  0%      4% 763 - hashcat-6.2.6/OpenCL/m11100_a3-optimized.cl                                                        8% 1209 - hashcat-6.2.6/OpenCL/m24100-pure.cl                                                15% 1568 - hashcat-6.2.6/modules/module_02000.so

**Reasoning**:
Now that the Hashcat archive is extracted, I need to find the exact path to the `hashcat` executable within the extracted directory to verify its installation.



In [9]:
LATEST_HASHCAT_VERSION = "6.2.6"
extracted_dir = f"hashcat-{LATEST_HASHCAT_VERSION}"
print(f"Listing contents of the extracted directory: {extracted_dir}")
!ls -l {extracted_dir}

Listing contents of the extracted directory: hashcat-6.2.6
total 4244
drwxr-xr-x 6 root root    4096 Sep  2  2022 charsets
drwxr-xr-x 3 root root    4096 Sep  2  2022 docs
-rw-r--r-- 1 root root      72 Sep  2  2022 example0.cmd
-rw-r--r-- 1 root root  214302 Sep  2  2022 example0.hash
-rwxr-xr-x 1 root root      66 Sep  2  2022 example0.sh
-rw-r--r-- 1 root root      63 Sep  2  2022 example400.cmd
-rw-r--r-- 1 root root      35 Sep  2  2022 example400.hash
-rwxr-xr-x 1 root root      56 Sep  2  2022 example400.sh
-rw-r--r-- 1 root root      56 Sep  2  2022 example500.cmd
-rw-r--r-- 1 root root      35 Sep  2  2022 example500.hash
-rwxr-xr-x 1 root root      50 Sep  2  2022 example500.sh
-rw-r--r-- 1 root root 1069601 Sep  2  2022 example.dict
drwxr-xr-x 3 root root    4096 Sep  2  2022 extra
-rwxr-xr-x 1 root root 1215952 Sep  2  2022 hashcat.bin
-rw-r--r-- 1 root root 1437184 Sep  2  2022 hashcat.exe
-rw-r--r-- 1 root root  240526 Sep  2  2022 hashcat.hcstat2
drwxr-xr-x 2 root root  

**Reasoning**:
Based on the `ls -l` output, `hashcat.bin` is the executable for Linux. I will now run `hashcat.bin -h` to verify the installation.



In [10]:
LATEST_HASHCAT_VERSION = "6.2.6"
extracted_dir = f"hashcat-{LATEST_HASHCAT_VERSION}"

# Construct the full path to the hashcat executable
hashcat_executable_path = f"./{extracted_dir}/hashcat.bin"

print(f"Verifying Hashcat installation with: {hashcat_executable_path} -h")
!{hashcat_executable_path} -h

Verifying Hashcat installation with: ./hashcat-6.2.6/hashcat.bin -h
hashcat (v6.2.6) starting in help mode

Usage: hashcat [options]... hash|hashfile|hccapxfile [dictionary|mask|directory]...

- [ Options ] -

 Options Short / Long           | Type | Description                                          | Example
================================+======+======================================================+=======================
 -m, --hash-type                | Num  | Hash-type, references below (otherwise autodetect)   | -m 1000
 -a, --attack-mode              | Num  | Attack-mode, see references below                    | -a 3
 -V, --version                  |      | Print version                                        |
 -h, --help                     |      | Print help                                           |
     --quiet                    |      | Suppress output                                      |
     --hex-charset              |      | Assume charset is given in hex   

## Final Task

### Subtask:
Hashcat'in başarıyla kurulduğunu ve kullanılabilir olduğunu onayla.


## Summary:

### Q&A
Yes, Hashcat has been successfully installed and confirmed to be usable. The final verification step, executing `hashcat.bin -h`, produced the expected help output, indicating that version 6.2.6 is functional.

### Data Analysis Key Findings
*   The previous `/content/hashcat` directory was successfully removed, ensuring a clean installation environment.
*   Initial attempts to download Hashcat version 6.2.6 in `.tar.gz` format from GitHub releases failed with a "404 Not Found" error, as pre-compiled Linux binaries were not available in this format.
*   Successfully downloaded the Hashcat version 6.2.6 pre-compiled Linux binary in `.7z` format.
*   The `p7zip-full` package was installed to enable extraction of the `.7z` archive.
*   Hashcat version 6.2.6 was successfully extracted into a directory named `hashcat-6.2.6`.
*   The `hashcat.bin` executable was identified within the extracted directory.
*   The installation was verified by executing `./hashcat-6.2.6/hashcat.bin -h`, which successfully displayed the help menu for Hashcat version 6.2.6.

### Insights or Next Steps
*   When downloading pre-compiled tools from GitHub releases, it's crucial to verify the correct archive format (e.g., `.7z` vs. `.tar.gz`) for the target operating system to avoid download failures.
*   The installed Hashcat executable is located at `./hashcat-6.2.6/hashcat.bin`; this path should be used for subsequent operations or added to the system's PATH variable for easier access.
